### PR TITLE
Make the file lookup faster

### PR DESCRIPTION
**Pull request by lrh2000**
This introduces `WithFileTable` and `get_file_fast` which allow lock-free operation on the file table and atomic-free file lookup respectively (if the file table is not shared, see #1550).
```rust
/// A helper trait that provides methods to operate the file table.
pub trait WithFileTable {
    /// Calls `f` with the file table.
    ///
    /// This method is lockless if the file table is not shared. Otherwise, `f` is called while
    /// holding the read lock on the file table.
    fn read_with<R>(&mut self, f: impl FnOnce(&FileTable) -> R) -> R;
}

impl WithFileTable for RwArc<FileTable> {
    fn read_with<R>(&mut self, f: impl FnOnce(&FileTable) -> R) -> R {
        if let Some(inner) = self.get() {
            f(inner)
        } else {
            f(&self.read())
        }
    }
}
```
Example:
https://github.com/lrh2000/asterinas/blob/a8f79962cf73b8645ea3a204f3b39f24fb5de6f9/kernel/src/syscall/ioctl.rs#L43-L47

```rust
/// Gets a file from a file descriptor as fast as possible.
///
/// If the file table is not shared with another thread, this macro will be free of locks
/// ([`RwArc::read`]) and free of reference counting ([`Arc::clone`]).
///
/// If the file table is shared, the read lock is taken, the file is cloned, and then the read lock
/// is released. This is because many operations on files can block, so we cannot hold such locks
/// when operating on files.
///
/// # Panics
///
/// This macro will cause an immediate panic if the [`RefCell`] that stores the file table in
/// [`ThreadLocal`] is already borrowed. Note that this also means that calling this macro twice
/// without releasing the [`RefMut`] of the file table will cause a panic.
///
/// [`RefCell`]: core::cell::RefCell
/// [`RefMut`]: core::cell::RefMut
/// [`ThreadLocal`]: crate::process::posix_thread::ThreadLocal
macro_rules! get_file_fast {
    { let ($file_table:ident, $file:ident) = $file_desc:ident @ $thread_local:expr } => {
        use alloc::borrow::Cow;

        let mut $file_table = $thread_local.file_table().borrow_mut();
        let $file = if let Some(file_table) = $file_table.get() {
            Cow::Borrowed(file_table.get_file($file_desc)?)
        } else {
            Cow::Owned($file_table.read().get_file($file_desc)?.clone())
        };
    };
}
```
Example:
https://github.com/lrh2000/asterinas/blob/a8f79962cf73b8645ea3a204f3b39f24fb5de6f9/kernel/src/syscall/accept.rs#L50-L51

### CODE REVIEW

**Comment by tatetian** [comment]
Great work! Introducing `RwArc` is a cool idea.

My only concern is the `get_file_fast` macro, which is not quite intuitive. For example, the original user code looks like this.

```rust
get_file_fast! { let (file_table, file) = sockfd @ ctx.thread_local };
```

I suppose that it is possible to remake the macro so that the user code would be more intuitive.

```rust
let file = get_file_fast!(ctx.thread_local, sockfd)?;
```

The trick here is to introduce a `Cow` type specialized for files.

```rust
/// A file object that is either owned or borrowed from the file table.
pub struct CowFile<'a> {
    Borrowed(&'a Arc<dyn FileLike>, &'a FileTable),
    Owned(Arc<dyn FileLike>),
}
```

With this helper type, the macro can be rewritten to the following form.

```rust
macro_rules! get_file_fast {
    ($thread_local:expr, $file_desc:expr) => {
        try {
            use crate::fs::file_table::CowFile;

            let mut file_table = $thread_local.file_table().borrow_mut();
            let file_desc = $file_desc:expr;

            // Fast path: when the file table is owned solely by this this thread, we gets the reference directly.
            let file = if let Some(file_table) = $file_table.get() {
                let file = file_table.get_file(file_desc);
                CowFile::from(file, file_table)
            }
            // Slow path: when the file table is shared by multiple threads, we have to lock and clone.
            else {
                let file = file_table.read().get_file(file_desc)?;
                CowFile::from(file.clone())
            }
            Ok(file)
        }
    }
}
```

**Comment by lrh2000** [comment]
@tatetian Thanks for your advice, but I don't think it will work:

```rust
                let file = file_table.get_file(file_desc);
                CowFile::from((file, file_table))
```
`file` is derived from the referenced `file_table`, how can we move `file_table` without triggering the compiler errors?

This is just like:
```rust
fn main() {
    let (_vec, _elem) = {
        let vec = vec![1, 2, 3];
        let elem = &vec[0];
        (vec, elem)
    };
}
```
which cannot compile:
```
error[E0505]: cannot move out of `vec` because it is borrowed
 --> x.rs:5:10
  |
3 |         let vec = vec![1, 2, 3];
  |             --- binding `vec` declared here
4 |         let elem = &vec[0];
  |                     --- borrow of `vec` occurs here
5 |         (vec, elem)
  |          ^^^  ---- borrow later used here
  |          |
  |          move out of `vec` occurs here
  |
help: consider cloning the value if the performance cost is acceptable
  |
4 |         let elem = &vec.clone()[0];
  |                        ++++++++
```

**Comment by tatetian** [comment]
Well.. the original proposal did not get the details right, but I still think it is a viable approach.

> cannot move out of XXX because it is borrowed

We need to get around this compiler error. Since we borrows a file from the file table, we must not move the file table. So it seems that the macro caller must prepare the file table first and pass it to the `get_file_fast` macro. Here is a PoC that I have tested on Rust playground.

```rust
use std::sync::Arc;
use std::cell::{RefCell, Ref, RefMut};

pub enum CowFile<'a> {
    Borrowed((&'a RefMut<'a, FileTable>, &'a Arc<File>)),
    Owned(Arc<File>),
}

pub struct ThreadLocal {
    file_table: RefCell<FileTable>
}

pub struct FileTable {
    single_file: Arc<File>,
}

pub struct File;

macro_rules! get_file_fast {
    ($file_table:expr, $file_desc:expr) => {
        {
            let file_table = $file_table;
            let file_desc = $file_desc;
            let file = &file_table.single_file;
            let cow_file = if true {
                CowFile::Borrowed((file_table, file))    
            } else {
                CowFile::Owned(file.clone())    
            };
            cow_file
        }
    }
}

fn main() {
    let thread_local = ThreadLocal {
        file_table: RefCell::new(FileTable { single_file: Arc::new(File) })
    };
    let file_table = thread_local.file_table.borrow_mut();
    let file = get_file_fast!(&file_table, 0);
}
```

Assuming that the approach shown by the PoC works on our real code, it would still be more complex than the original one as it requires two lines of code, one for the file table, the other for the file.

The line to acquire file table can be hidden if we change the interface of the macro so that the caller gets the file within a closure.

```rust
get_file_fast!(thread_local, fd, |file| {
   ...
})
```

Using this macro will require a larger refactoring work on the user side. For example, the `sys_fsync` syscall will be changed from

```rust
pub fn sys_fsync(fd: FileDesc, ctx: &Context) -> Result<SyscallReturn> {
    debug!("fd = {}", fd);
    get_file_fast! { let (file_table, file) = fd @ ctx.thread_local };
    let dentry = file.as_inode_or_err()?.dentry();
    dentry.sync_all()?;
    Ok(SyscallReturn::Return(0))
}
```

to

```rust
pub fn sys_fsync(fd: FileDesc, ctx: &Context) -> Result<SyscallReturn> {
    debug!("fd = {}", fd);
    get_file_fast!(ctx.thread_local, fd, |file| {
        let dentry = file.as_inode_or_err()?.dentry();
        dentry.sync_all()?
        Ok(SyscallReturn::Return(0))
    })
}
```

One particularly good thing about this new form of `get_file_fast` is that it completely hides the implementation details of acquiring the file table. Most users of `get_file_fast` do not have any interest in accessing the file table.

**Comment by lrh2000** [comment]
> Using this macro will require a larger refactoring work on the user side. For example, the `sys_fsync` syscall will be changed from
> 
> ```rust
> pub fn sys_fsync(fd: FileDesc, ctx: &Context) -> Result<SyscallReturn> {
>     debug!("fd = {}", fd);
>     get_file_fast! { let (file_table, file) = fd @ ctx.thread_local };
>     let dentry = file.as_inode_or_err()?.dentry();
>     dentry.sync_all()?;
>     Ok(SyscallReturn::Return(0))
> }
> ```
> 
> to
> 
> ```rust
> pub fn sys_fsync(fd: FileDesc, ctx: &Context) -> Result<SyscallReturn> {
>     debug!("fd = {}", fd);
>     get_file_fast!(ctx.thread_local, fd, |file| {
>         let dentry = file.as_inode_or_err()?.dentry();
>         dentry.sync_all()?
>         Ok(SyscallReturn::Return(0))
>     })
> }
> ```
> 
> This looks acceptable to me.

This particular example may look acceptable, but I am not very comfortable with this approach because there are too many system calls trying to get files. Given the number of usage sites, I'd like to provide a non-closure-based implementation so that users can use it easily.

In your particular example, the system call is only a few lines, so it looks good, but the problem is that some system calls are much longer than this.
https://github.com/asterinas/asterinas/blob/87ee88da8c5e6297838a86b395cee3c56807df60/kernel/src/syscall/pwritev.rs#L47-L107

We can try to refactor such system calls, but it won't change the fact that closure-based `get_file_fast` is generally harder to use.

> Assuming that the approach shown by the PoC works on our real code, it would still be more complex than the original one as it requires two lines of code, one for the file table, the other for the file.

Is this acceptable to you? I think it's somehow better than switching to the closure-based approach.

Meanwhile, I was wondering if you had any ideas about revising the syntax of the original macro (if that's the only concern)? I don't really have any good ideas, just some possibilities.
```rust
// get the file "at" (@) the file table
get_file_fast! { let (file_table, file) = sockfd @ ctx.thread_local };

// similar to "match xxx { yyy @ zzz => { .. }}"
get_file_fast! { (ctx.thread_local, sockfd) @ let (file_table, file) };

// use "<="
get_file_fast! { let (file_table, file) <= (ctx.thread_local, sockfd) };

// any other ideas?
```

**Comment by tatetian** [comment]
Let's go through the three candidate syntaxes of `get_file_fast` one by one.

First of all, I have no problem with a two-liner. The length is not the problem. In fact, being more explicit and clear (the first line acquires the file table, the second one gets the file) makes it the perfect syntax. My problem with it is about the mutability.

```rust
    let mut file_table = ctx.thread_local.file_table().borrow_mut();
    let file = get_file_fast!(&mut file_table, fd);
```

Put yourself in the shoes of a reader: if you don't know the details about `get_file_fast` (using `RwArc::get` internally), you would have no clue why you need to call `borrow_mut` -- you are not intended to update the file table.

Second, thanks for trying out new variants with the original macro. But I still don't think they are ideal. They look alien to the Rust language. The syntax is just too mysterious. And this macro's implicitly defining a variable for file table is simply meant to satisfy the Rust compiler (correct me if I was wrong).

```rust
// get the file "at" (@) the file table
get_file_fast! { let (file_table, file) = sockfd @ ctx.thread_local };

// similar to "match xxx { yyy @ zzz => { .. }}"
get_file_fast! { (ctx.thread_local, sockfd) @ let (file_table, file) };

// use "<="
get_file_fast! { let (file_table, file) <= (ctx.thread_local, sockfd) };
```

Lastly, let's get back to the closure-based syntax. Its syntax is very intuitive. It does not introduce a useless file table variable simply to make the compiler happy. It even prevents you from making mistakes. A user may attempt to write such _natural_ but _unsupported_ usage code.

```rust
get_file_fast! { let (file_table, file) = fd1 @ ctx.thread_local };
get_file_fast! { let (file_table, file) = fd2 @ ctx.thread_local };
```

or 

```rust
    let mut file_table = thread_local.file_table.borrow_mut();
    let file1 = get_file_fast!(&mut file_table, fd1);
    let file1 = get_file_fast!(&mut file_table, fd2);
```

But the user would be be much more uncomfortable with this.

```rust
get_file_fast!(thread_local, fd1, |file| {
    get_file_fast!(thread_local, fd2, |file| {
       ...
    })
})
```

> it won't change the fact that closure-based get_file_fast is generally harder to use.

Sure. Closure is harder to use. For the particular example you gave, the refactored code would look like this.

```rust
 fn do_sys_pwritev( 
     fd: FileDesc, 
     io_vec_ptr: Vaddr, 
     io_vec_count: usize, 
     offset: i64, 
     _flags: RWFFlag, 
     ctx: &Context, 
 ) -> Result<usize> { 
     // TODO: Implement flags support 
     debug!( 
         "fd = {}, io_vec_ptr = 0x{:x}, io_vec_counter = 0x{:x}, offset = 0x{:x}", 
         fd, io_vec_ptr, io_vec_count, offset 
     ); 
     if offset < 0 { 
         return_errno_with_message!(Errno::EINVAL, "offset cannot be negative"); 
     }

     get_file_fast!(ctx.thread_local, fd, |file| {
        _do_sys_pwritev(
            file,
             io_vec_ptr: Vaddr, 
             io_vec_count: usize, 
             offset: i64, 
             _flags: RWFFlag, 
             ctx: &Context, 
        )
     })
}

fn _do_sys_pwritev( 
     file: &Arc<File<dyn FileLike>>, 
     io_vec_ptr: Vaddr, 
     io_vec_count: usize, 
     offset: i64, 
     _flags: RWFFlag, 
     ctx: &Context, 
) -> Result<usize> {      
     // TODO: Check (f.file->f_mode & FMODE_PREAD); We don't have f_mode in our FileLike trait 
     if io_vec_count == 0 { 
         return Ok(0); 
     } 
  
     let mut total_len: usize = 0; 
     let mut cur_offset = offset as usize; 
  
     let mut reader_array = VmReaderArray::from_user_io_vecs(ctx, io_vec_ptr, io_vec_count)?; 
     for reader in reader_array.readers_mut() { 
         if !reader.has_remain() { 
             continue; 
         } 
  
         let reader_len = reader.remain(); 
         if total_len.checked_add(reader_len).is_none() 
             || total_len 
                 .checked_add(reader_len) 
                 .and_then(|sum| sum.checked_add(cur_offset)) 
                 .is_none() 
             || total_len 
                 .checked_add(reader_len) 
                 .and_then(|sum| sum.checked_add(cur_offset)) 
                 .map(|sum| sum > isize::MAX as usize) 
                 .unwrap_or(false) 
         { 
             return_errno_with_message!(Errno::EINVAL, "Total length overflow"); 
         } 
  
         // TODO: According to the man page 
         // at <https://man7.org/linux/man-pages/man2/readv.2.html>, 
         // writev must be atomic, 
         // but the current implementation does not ensure atomicity. 
         // A suitable fix would be to add a `writev` method for the `FileLike` trait, 
         // allowing each subsystem to implement atomicity. 
         let write_len = file.write_at(cur_offset, reader)?; 
         total_len += write_len; 
         cur_offset += write_len; 
     } 
     Ok(total_len) 
 } 
```

This single example still looks alright, although I can't make a conclusion that all usage are acceptable as I didn't examined all use cases.

To sum up, I think the first option is the perfect if the mutability issues can be avoided. The second option is just unacceptable to my taste (everyone has a say on this; just my opinion). The third option is a compromised solution. It is ok to me, but I am not enthusiastic about it.

**Comment by lrh2000** [comment]
@tatetian Thanks for the detailed comments.

> My problem with it is about the mutability.
> 
> ```rust
>     let mut file_table = ctx.thread_local.file_table().borrow_mut();
>     let file = get_file_fast!(&mut file_table, fd);
> ```
> 
> Put yourself in the shoes of a reader: if you don't know the details about `get_file_fast` (using `RwArc::get` internally), you would have no clue why you need to call `borrow_mut` -- you are not intended to update the file table.

IMO, it's not a big deal. "Mutability" is a bad name that cannot be fixed due to widespread usage and Rust's stability guarantee. However, any qualified Rust programmer should be aware that "mutable reference" really means "exclusive reference" and "immutable reference" really means "shared reference" -- so holding a mutable reference does not mean we want to mutate the data, it means we want exclusivity.
 
> And this macro's implicitly defining a variable for file table is simply meant to satisfy the Rust compiler (correct me if I was wrong).

Well, from my perspective, it's a helper macro to avoid writing `borrow_mut` to the file table repeatedly. As a helper macro, I assume that the user needs to be aware that it's a `RefMut` of the file table, so the user can drop it if necessary.
 - By this I mean that even if there is a syntax (e.g., there is already a [pre-RFC that introduces `super let`](https://hackmd.io/@rust-lang-team/HJpRzPzoa?utm_source=preview-mode&utm_medium=rec)) that can avoid defining this `file_table` variable, we may still want to keep it. Otherwise, the user will not be able to manually drop it. So when it borrows the file table again, there will be a panic.

If you disagree with consiseness (e.g., avoid repeating `borrow_mut`) and instead prefer explicitness (e.g., repeat `borrow_mut`), let's abandon this approach and focus on the other two.

> It even prevents you from making mistakes. A user may attempt to write such _natural_ but _unsupported_ usage code.
> 
> ```rust
> get_file_fast! { let (file_table, file) = fd1 @ ctx.thread_local };
> get_file_fast! { let (file_table, file) = fd2 @ ctx.thread_local };
> ```

I agree this is bad, however:

> or
> 
> ```rust
>     let mut file_table = thread_local.file_table.borrow_mut();
>     let file1 = get_file_fast!(&mut file_table, fd1);
>     let file1 = get_file_fast!(&mut file_table, fd2);
> ```

I don't think it's a big problem, because:
 1. If the user does not use `file1` after the second `get_file_fast!`, it will compile and run without problems.
 2. If the user does use `file1` after the second `get_file_fast!`, this cannot even compile. Since the program is rejected at compile time (the user doesn't even have to run it!), I think this is perfectly fine.

> Sure. Closure is harder to use. For the particular example you gave, the refactored code would look like this.
> 
> ```rust
>  fn do_sys_pwritev( 
>      fd: FileDesc, 
>      io_vec_ptr: Vaddr, 
>      io_vec_count: usize, 
>      offset: i64, 
>      _flags: RWFFlag, 
>      ctx: &Context, 
>  ) -> Result<usize> { 
>      // TODO: Implement flags support 
>      debug!( 
>          "fd = {}, io_vec_ptr = 0x{:x}, io_vec_counter = 0x{:x}, offset = 0x{:x}", 
>          fd, io_vec_ptr, io_vec_count, offset 
>      ); 
>      if offset < 0 { 
>          return_errno_with_message!(Errno::EINVAL, "offset cannot be negative"); 
>      }
> 
>      get_file_fast!(ctx.thread_local, fd, |file| {
>         _do_sys_pwritev(
>             file,
>              io_vec_ptr: Vaddr, 
>              io_vec_count: usize, 
>              offset: i64, 
>              _flags: RWFFlag, 
>              ctx: &Context, 
>         )
>      })
> }
> 
> fn _do_sys_pwritev( 
>      file: &Arc<File<dyn FileLike>>, 
>      io_vec_ptr: Vaddr, 
>      io_vec_count: usize, 
>      offset: i64, 
>      _flags: RWFFlag, 
>      ctx: &Context, 
> ) -> Result<usize> {      
>      // TODO: Check (f.file->f_mode & FMODE_PREAD); We don't have f_mode in our FileLike trait 
>      if io_vec_count == 0 { 
>          return Ok(0); 
>      } 
>   
>      let mut total_len: usize = 0; 
>      let mut cur_offset = offset as usize; 
>   
>      let mut reader_array = VmReaderArray::from_user_io_vecs(ctx, io_vec_ptr, io_vec_count)?; 
>      for reader in reader_array.readers_mut() { 
>          if !reader.has_remain() { 
>              continue; 
>          } 
>   
>          let reader_len = reader.remain(); 
>          if total_len.checked_add(reader_len).is_none() 
>              || total_len 
>                  .checked_add(reader_len) 
>                  .and_then(|sum| sum.checked_add(cur_offset)) 
>                  .is_none() 
>              || total_len 
>                  .checked_add(reader_len) 
>                  .and_then(|sum| sum.checked_add(cur_offset)) 
>                  .map(|sum| sum > isize::MAX as usize) 
>                  .unwrap_or(false) 
>          { 
>              return_errno_with_message!(Errno::EINVAL, "Total length overflow"); 
>          } 
>   
>          // TODO: According to the man page 
>          // at <https://man7.org/linux/man-pages/man2/readv.2.html>, 
>          // writev must be atomic, 
>          // but the current implementation does not ensure atomicity. 
>          // A suitable fix would be to add a `writev` method for the `FileLike` trait, 
>          // allowing each subsystem to implement atomicity. 
>          let write_len = file.write_at(cur_offset, reader)?; 
>          total_len += write_len; 
>          cur_offset += write_len; 
>      } 
>      Ok(total_len) 
>  } 
> ```
> 
> This single example still looks alright, although I can't make a conclusion that all usage are acceptable as I didn't examined all use cases.

So we have to define another function and pass all the arguments to it, which is kind of annoying.

> To sum up, I think the first option is the perfect if the mutability issues can be avoided.

Not sure if my argument above solves your concern about "the mutability issues"? For example, we can note in the `get_file_fast` doc comments that we need the mutable reference because we want exclusivity.

**Comment by tatetian** [comment]
Then I suppose that our best option is the first approach.

```rust
    let mut file_table = ctx.thread_local.file_table().borrow_mut();
    let file = get_file_fast!(&mut file_table, fd);
```

Let's go with it 😄

**Review by tatetian — APPROVED** [review]
LGTM. Thanks for the contribution!
